### ISSUE TITLE
Strange reportPossiblyUnboundVariable

### ISSUE THREAD

**Issue by rikardn** [microsoft/pyright#8489]
When type checking this program:

```python
def func(t):
    if t == "BIRD":
        myvar = 23
    elif t == "PLANE":
        othervar = 98
    else:
        raise ValueError("Bad t")

    if t == "BIRD":
        print(myvar)
    elif t == "PLANE":
        print(othervar)
```

I get 

```
10:15 - error: "myvar" is possibly unbound (reportPossiblyUnboundVariable)
12:15 - error: "othervar" is possibly unbound (reportPossiblyUnboundVariable)
```

but can't we be certain that these variables are bound?

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

The "unbound variable" check is based on static code flow analysis. In your code, the type of `myvar` is `Literal[23]` in one branch, but it is "unbound" in other branches. Static type analysis doesn't track conditions like "the type of `myvar` is `Literal[23]` if `t` happened to be `"BIRD"` earlier in the program but is "unbound" otherwise. 

You have a few options. 
1. You can globally disable the `reportPossiblyUnboundVariable` check in your project if you want to write code that depends on this behavior.
2. You can use `# type: ignore` or `# pyright: ignore` to suppress individual instances of this diagnostic check.
3. (my personal preference) Don't write code that initializes symbols on some branches of your code and not on others. Restructure your code to avoid this entirely or preinitialize variables with `None` or some other relevant default value like `0` in the case of an integer.

**Comment by rikardn** [comment]
Thanks for your explanation. My interpretation is that although it would be possible to implement a check for this it would be a bad idea because the error from pyright actually tells us that the code could be made better for example via your option 3.

**Comment by erictraut** [comment]
It is theoretically possible to implement a check like this, but it would be infeasible for real-world code because the complexity would explode combinatorially. I don't know of any static type checkers or compilers in any language that perform this sort of "conditional type tracking" for this reason.
